### DISCUSSION TITLE
useSWR + Suspense leads to infinite rerenders

### DISCUSSION THREAD

**Discussion by DevAtHeart** [vercel/swr#1109] [Q&A]
Hello,

I have a problem that some configuration of useSWR together with Suspense lead to infinite rerenders but some do not.

I use:
```
react@17.0.2
next@10.1.3
swr@0.5.5
```

I wrap my components in:
```javascript
export const WidgetData: React.FC<WidgetDataProps> = ({ fallback, children }) => {
  if (isServer) return <>{fallback}</>
  return (
    <WidgetErrorBoundary>
      <Suspense fallback={fallback}>{children}</Suspense>
    </WidgetErrorBoundary>
  )
}
```

It is used like:
```javascript
<WidgetData fallback={<>Loading...</>}>
  <MyWidget query={query} />
</WidgetData>
```

`MyWidget` looks like:

```javascript
export const MyWidget: React.FC<{query: string}> = ({query}) => {
  const { data } = useSWR([`/api/data?query=${query}`, query], fetcher)
  
  return <>{data}</>
}

```

It will never render data but keep showing the fallback. In the background, the http request is repeated forever.

With suspense false, everything works perfectly.
With suspense true, it starts randomly going into infinite render loops, maybe a race condition?
It doesn't happen for all widgets, for most it NEVER happens.
It also never happens on first render.
It happens in dev mode and production.

**Comment by DevAtHeart** [comment]
I managed to reproduce it here: https://codesandbox.io/s/swr-suspense-bug-f5owj

Just enter something into the input and check console.

I think it has to do with the request returning the same result as the cache (previous result). You can use the commented line to fix the issue! I want to verify,  I am not doing something wrong before opening an issue..

**Reply by icyJoseph** [reply]
Fixed by https://github.com/vercel/swr/pull/1058 - which is not released yet. 

If you switch to https://www.npmjs.com/package/swr/v/1.0.0-beta.5 the issue is fixed there.

**Reply by DevAtHeart** [reply]
I tried using beta but had another issue regarding #339 thinking it didn't work! Thank you so much!

Apropos #339: Is it recommended to not use conditional fetching and instead handle the use-case via early return in the fetcher for instance?

**Reply by shuding** [reply]
I don't think #339 will have a proper solution (suspense **and** conditional fetching), because suspense is not a good fit for conditional fetching. It has to either _throw_ or _render_ (also suspense is still experimental). I think an early return and tackle it manually would be a work around for now.
